staging/src/k8s.io/apiserver/pkg/authentication/authenticator/interfaces.go

```diff
@@ -77,4 +78,37 @@ type Response struct {
 	Audiences Audiences
 	// User is the UserInfo associated with the authentication context.
 	User user.Info
+	// AuthenticatorName identifies the authenticator that authenticated the request.
+	AuthenticatorName string
+}
+
+// Error is the error returned by an authenticator upon failures.
+type Error struct {
+	// Err is the error produced by an authenticator.
+	Err error
+	// AuthenticatorName identifies the authenticator that failed an authentication request.
+	AuthenticatorName string
+}
+
+// Error implements the errors.Error interface.
+func (a *Error) Error() string {
+	return fmt.Sprintf("authenticator: %q failed to process authentication request: %v", a.AuthenticatorName, a.Err)
+}
+
+// Unwrap unpacks the wrapped error.
+func (a *Error) Unwrap() error {
+	if a.Err == nil {
+		return nil
+	}
+
+	return a.Err
+}
+
+// NewError constructs Error.
+// When the supplied error is nil, this method will also return nil.
+func NewError(authenticatorID string, err error) *Error {
```

Comment:
This needs to return `error` otherwise it will result in a typed `nil`.

You could also remove the special handling of `nil`.

Reply:
Done.